README
See also: http://riemann.fmi.uni-sofia.bg/ninit/

Each service gets its own directory under /etc/ninit (change this in the
source, it's a #define right at the start of ninitfeatures.h).


Each service directory can contain the following files/symlinks:

  depends

    a plain text file containing a service name per line.
    Example: /etc/ninit/sshd/depends could contain "network".
    Each of these services will be started before this service is
    started.  If you need to wait for static initializations to
    complete, use the sync flag.  See also the wait flag below.

  run

    a symbolic link to the program name.  No hard link, because argv[0]
    for the programs is created by extracting the part after the last
    slash in the contents of the symbolic link.
    Example: "/usr/bin/sshd" would be run with argv[0]="sshd".

  end

    similar to run.  After the service finish "end" will be executed.
    The helper does not set environ, softlimit, uid/gid restrictions.
    You can put there for example:
	#!/bin/sh
	exec /bin/nsvc -o service_name
    Use this option instead of sync or wait mode.  Let for example
    services B,C,D must be started after A finish.  There are two
    possible solutions:
    Bad:  start A in sync mode
    Good: start A i normal mode and put in "end"
	#!/bin/sh
	exec /bin/nsvc -o B C D

  params

    a plain text file containing command line parameters for the
    service, one parameter per line.  No shell expansion is done.  If
    you need shell expansion, have run point to a shell script instead
    of the real daemon.  Note: Have the shell script exec the daemon
    instead of simply running it to save system resources.

  environ

    similar to params.  run appends contents of this file to the environ.
    If the first line has zero length, unset all environ.
    If a line does not have '=' (HOME), unset this variable.

  wait

    similar to params.  /etc/inint/run waits for services included in 
    this file to finish and after that start a current service.
    Include wait services also in depends.  This is different from sync.
    Let a service B must be started after a service A finish.
    One can sync A.  This will block all other services.
    In this case it's better to use wait instead of sync.
    It's possible to set wait parameters (see maxwait) at the end of
    each entry.  Example: some_service:180:3

  maxwait

    a plain text file containing the values n1:n2.  Wait no more 
    than n1 sec a service for finish.  See wait.  Default is 600 sec.  
    The value n1=0 means -- wait until the service finish.


  softlimit
 
    see DJB program softlimit.  Similar to params.  Lines
    with the same options as softlimit (skip leading '-').  Example:
    echo m300000 > softlimit; echo o30 >> softlimit

  respawn

    touch this file to make ninit respawn the process when it dies.
    This should be touched for getty and network servers.

  sync

    touch this file to make ninit wait until the service ends.  sync is
    mutually exclusive with respawn.  This is meant for static
    initializations like "ifconfig".  See also wait.
    If sync contains a nonzero number n ninit will wait max n sec
    the service to finish.  If it does not ends until n secs then 
    ninit continues to work.  Example: echo 600 > sync

  log

    if this directory exists, it is taken as service and ninit creates
    a named pipe log/in and soft link of log/in to out.  If the log 
    service can not be started, this service will block if it writes 
    to stdout.

  nice

    a plain text file containing the value to add to the nice level
    via the nice system call.

  sleep

    a plain text file containing the value n.  sleep n secs after fork
    before running child.	

  in

    this file(named pipe) is used for stdin.

  out

    this file(named pipe) is used for stdout and stderr.
    e.g. a symlink to /dev/null

  gid 
    
    a plain text file containing a number.
    Exec the program with this GID.  Example echo -n 234 > gid
    To activate this feature edit the file ninitfeatures.h 
    and install the package again.

  uid 
    
    exec the program with this UID.  It can contain also UID:GID
    It is possible to write here also the supplementary groups.
    For example echo 23:99:240:320:100 > uid start the service with:
    UID=23, GID=99, GROUPS=99,240,320,100.

  setup

    if this file exist it is started just before run in sync mode.
    If uid/gid or softlimit exist they are applied first and 
    after that setup is executed.
  
  rsetup

    similar to setup.  It is executed before setup as root and
    not softlimit, uid/gid, in/out restrictions.  If you
    want to create/modify params, wait, depends... see sys-rsetup flag.

  sys-rsetup

    similar to setup.  It is executed first as root.
    It can be used to create/modify params, wait, depends...

  pidfile

    easy method to setup services which run in background.
    Prepare them as ordinary services and write the name of the pidfile.
    Don't use link here!
    Example: echo -n /var/run/gpm.pid > pidfile
    
  alarm

    a plain text file containing a number n.
    Do alarm(n) before starting the child.

  pause

    if this file exist run selfkills with SIGSTOP signal.

  pause-wait

    if this file exist run-wait selfkills with SIGSTOP signal.

  cron

    a plain text file containing lines with numbers a:b[:c]
    for each line run wait find a solution of 
	cron = 60 * (a*x+b) > now  (x is integer)
    The smallest cron is next cron-start for this service.
    The number a=10080 (24*60*7) is special.  For example 10080:25
    means: Sunday 00:25:00 UTC

ninit will try to run the command line arguments as services.  The
kernel passes its arguments to init.  That means you can for example
have a service /etc/ninit/sos-shell that starts a /bin/sh and then use
LILO to boot "bzImage sos-shell".  ninit will then run that service.
If none of the services worked (or none were given), ninit will spawn
the service "default".  The normal way to configure ninit is to have
default be an empty service and just list all the services you want
ninit to start at boot time in default/depends.

Other special services (besides "default") are "ctrlaltdel" and "kbreq".
ctrlaltdel will be run when the console user presses ctrl-alt-del and is
meant to reboot the computer.  kbreq is the keyboard request, which
can be mapped using loadkeys.  On my box, it is on Alt+Arrow-Up.  I use
it to shut down the computer.

--------------------------
Memory allocation in NINIT.  ninit uses only alloca for RAM.

One service = 21 + strlen(service_name).  Example getty/1 is 28 bytes.
If you need to load more than 50 services then use at boot process 
something like:
	/sbin/ninit -M3200

after reload/reboot ninit will work with 3200 bytes buffer.
default buffer is 1536.  Good for 50-60 services.
If someone uses more than 50 services let me know.

--------------------------
After building a new ninit test/install it with:
	ninit-reload -v -u /path/to/new/ninit [-M2000]
	cp /path/to/new/ninit /sbin/ninit
	ninit-reload -v -u /sbin/ninit [argv1 argv2 ...]
some of argvs can be -M2000 or -M3200 ...

There is no need of reboot!  Dump memory buffer with:
	ninit-reload -m > /tmp/ninit.mem
	less /tmp/ninit.mem


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


reload.c
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/poll.h>
#include <limits.h>
#include "ninit.h"

#define USAGE "Usage:\n" \
"\tninit-reload ","-v","\n" \
"\tninit-reload ","-m > dump_memory_file\n" \
"\tninit-reload ","-d > dump_data_file\n" \
"\tninit-reload ","[Options] /path/to/ninit [arg(s)...]\n", \
"\nOptions:\n\t[-t time_diff]\t[-f data_file]\t" \
"[-u]\t\t[-v]\n\t" \
"[-R service]\t[-r number]\t[-a auto-size]\n\t" \
"[-E file]\t[-e var=name]\t[-e variable]\t[-e -]\n"

#define UNKNOWN_PROCESS_SIZE

static char *buf, *flag_file;
static unsigned int verbose;
static char do_update, get_memory, get_data;

#define UL sizeof(uint32t)

void die_xx(char *s) { errmsg_iam("ninit-reload"); err(2,s,0); _exit(1); }

static int safe_read_code(char code) {
  int r, k;
  for (k=0; k<50; k++) {
    r = read(outfd,buf,BUFFER_TMP_LEN);
    if (r == 1 && buf[0] == code) return 0;
    if (r <= 0) nano_sleep(0, 20000000); /* 0.020 sec */
    else break;
  }
  return -1;
}
static void safe_read() { 
  if (safe_read_code('1')) die_xx("wrong responce from ninit");
}
static void safe_read_execve(char code, char *s) {
  if (safe_read_code(code) == 0) return;
  die(1, "Ops! I got exec error from ",s);
}
static void safe_write(int len) {
   if (len != write(infd, buf, len)) die_xx("error writing to ninit");
}
static int safe_cat(char *dest, char *src, int len) {
  int k = str_len(dest);
  if (k<len) k += str_copyn(dest+k, src, len-k);
  dest[k] = 0;
  return k;
}
static void put_arg(char *string) {
  uint32t *len = (uint32t *)buf;
  buf[UL] = 0;
  *len = safe_cat(buf+UL, string, 300);
  safe_write(*len + UL);
  safe_read();
}

#include "check_opt.h"
#include "get_services.h"
#include "open_inout.h"

static char *do_AutoSize(uint32t len) {
  uint32t u;
  char *x;
  if (len>300) len=300;
  u = (maxprocess + len + 8) * (PROCESS_SIZE + 12);

  x = fu(u & ~15);
  *--x = 'M';  *--x = '-';

#ifndef INIT_TIMEOUT_WRITE
  if (u > PIPE_BUF - 256)
    carp("\n\a*** WARNING *** for option ",x, " ninit must be build\n"
	 "with INIT_TIMEOUT_WRITE flag.  See ninitfeatures.h");
#endif
  return x;
}

static char *align_name(char *s) {
  static char x[20];
  int len = str_len(s);
  byte_set(x,sizeof(x),' ');
  if (len<=16) len=18 - len;
  else { len=1; x[0] = '\t'; }
  x[len] = 0;
  return x;
}

int main(int argc, char **Argv) {
  char *remove_P[argc], *env_c[argc], **env=0;
  char *auto_memory=0, **argv;
  time_t TimeDiff=0;
  int do_Remove=0, do_remove=0, remove_p[argc], AutoSize=0;
  int i,j,k,env_len=0;
  uint32t *len;
  char *initroot, *x;
 
  INIT_ENV_GET_HOME(initroot,"NINIT_HOME","INIT_HOME");
  buf = alloca(BUFFER_TMP_LEN);
  len = (uint32t *)buf;

  mem.a = INIT_ALLOC_BUFFER;
  if (argc < 2) { usage: die(1, USAGE); }
  if (initroot == 0) initroot = INITROOT;
  argv=Argv;

  argv++;
  errmsg_iam("ninit-reload");
  while (argv[0] && argv[0][0]=='-') {
    switch(argv[0][1]) {
    case 'v': verbose++; break;
    case 'u': do_update=1; break;
    case 'm': get_memory=1; break;
    case 'd': get_data=1; break;
    case 'f': chk_opt(argv,x); flag_file = x; break;
    case 't': chk_opt(argv,x); TimeDiff = x_atoi(x); break;
    case 'a': chk_opt(argv,x); AutoSize = x_atoi(x); break;
    case 'R': chk_opt(argv,x); remove_P[do_Remove++] = x; break;
    case 'e': chk_opt(argv,x);
      /* next clears all like env -i!  to remove '-' do; 
       * echo > /tmp/x; ninit-raload -E /tmp/x ... */
      if (x[0]=='-' && x[1]==0) x=""; 
      env_c[env_len++] = x; break;
    case 'E': chk_opt(argv,x);
      i=open(x, O_RDONLY);
      if (i<0) die(1, "unable to open ", x);
      else {
	if (GLOBAL_READ(i,x, j,12000)) _exit(1);
	close(i);
	x[j] = 0;

	j = splitmem(0,x,'\n');
	env = alloca((argc + j + 5) * sizeof(char *));
	if (env == 0) _exit(1);
	splitmem(env,x,'\n');
	skip_comments(env);
      }
      break;
    case 'r':
      chk_opt(argv,x);
      if ('0'<=x[0] && x[0]<='9') remove_p[do_remove++] = x_atoi(x);
      break;
    default:
      die(1, "Unknown Option: ",argv[0]);
    }
    argv++;
  }

  argc -= (argv-Argv);

  x = env_get("NINIT_MEMORY");
  if (x) mem.a = atoulong(x);
  mem.r=mem.a;
  if (mem.a==0 || mem.a > 64000 || (mem.x=alloca(mem.a))==0) _exit(1);

  if (env==0) env=env_c;
  else {
    char **xx=env+1, *s;
    for (k=1; (s=*xx); xx++) if (*s) env[k++] = s;
    byte_copy(env+k, env_len * sizeof(char *), env_c);
    env_len += k; 
  }

  if (flag_file==0) open_inout(initroot);
  else {
    if (*flag_file == 0) goto usage;
    infd=-1;
    outfd=open(flag_file, O_RDONLY);
    if (outfd<0) die(1, "could not open ",flag_file);
    process_size = lseek(outfd,0,SEEK_END);
    lseek(outfd,0,SEEK_SET);
  } 

  errmsg_iam(0);
  if (get_memory || get_data) {
    struct pollfd pfd;
    pfd.fd=outfd;
    pfd.events=POLLIN;
    buf[0] = 'D';
    buf[1] = (get_memory) ? 'M' : '1';
    write(infd,buf,2);

    for (;;) {
      j=poll(&pfd,1,400);
      if (j==-1) {
	if (errno==EINTR) { carp("interrupt signal"); }
    	else die_xx("poll fialed!");
      } else if (j==1) {
	i=read(outfd,buf,BUFFER_TMP_LEN);
	if (i>0) write(1,buf,i);
      } else {
	return 0;
      }
    }
  }
  
  get_services();

  if (verbose) {
    char *what;
    msg("services from: ",
	(flag_file) ? "" : initroot, 
	(flag_file) ? flag_file : "/out");
    for (j=0; j<= maxprocess; j++) {
      switch (root[j].pid) {
      case 1: what = "finish"; break;
      case 0: what = "down"; break;
      default: what = fu(root[j].pid);
      }
      x = root_name(j);
      msg(fu(j),".\t",x,align_name(x),what);
    }
    if (!do_update) return 0;
  }

  if (do_update) {
    if (!argv || !argv[0] || argv[0][0]!='/')
      die_xx("I need the full /path/of/ninit/program");
    
    k = str_len(argv[0]);
    if (k > 240) die_xx("too long file name.  must be <= 240.");
    
    if (access(argv[0], X_OK))
      die(1, "Ops! Check the mode of ", argv[0]);

    if (AutoSize) auto_memory=do_AutoSize(AutoSize);
    
    errmsg_puts(1,"\nreplacing current ninit with: ");
    fmt_argv(1,argv, " ");
    msg("",auto_memory);

    if (flag_file) {
      close(infd); close(outfd);
      open_inout(initroot);
    }

    buf[0]='U'; buf[1]='U'; buf[2]=0;

    if (argc>1 || auto_memory || env_len) {
      /* ----- to ninit ----- */
      safe_cat(buf, initroot, 200);
      k = safe_cat(buf, "/sys/update", 230);
      if (access(buf+2, X_OK))
	die(1, "Ops! Check the mode of ", buf+2);

      buf[1] = 0;
      buf[k++] = 0;
      if (verbose) msg("sending reload signal to ","ninit");
      safe_write(k);
      
      /* ----- from/to update ----- */
      safe_read_execve('7', "~/sys/update");
      *len=argc + 1 + (auto_memory != 0);
      safe_write(UL);
      safe_read();
      
      for (i=0; i<argc; i++) put_arg(argv[i]);	/* argv */
      if (auto_memory) put_arg(auto_memory);
      put_arg("");
      
      *len=env_len;
      safe_write(UL);
      safe_read();
      for (i=0; i<env_len; i++) put_arg(env[i]);
      
      *len = 1953066852;
      if (verbose) msg("sending reload signal to ","~/sys/update","\n");
      safe_write(UL);
    } else {
      k=safe_cat(buf, argv[0], 300);
      buf[1] = 0;
      buf[k++] = 0;
      if (verbose) msg("sending reload signal to ","ninit","\n");
      safe_write(k);	/* execve argv[0] */
    }      

    /* ----- from/to ninit ----- */
    safe_read_execve('8',argv[0]);
    if (verbose) msg("now trying to restore services:");
    
    for (j=0; j<= maxprocess; j++) {
      char *name=root_name(j);

      for (k=0; k<do_remove; k++)
	if (j==remove_p[k]) {
	  if (verbose) { msg("\t",name,align_name(name),"removed"); }
	  k=-1; break;
	}
      if (k==-1) continue;

      for (k=0; k<do_Remove; k++)
	if (!str_diff(remove_P[k],name)) {
	  if (verbose) { msg("\t",name,align_name(name),"removed"); }
	  k=-1; break;
	}
      if (k==-1) continue;

      root[j].pr_circular = 0;
      root[j].startedat += TimeDiff;
      buf[0]='U'; buf[1]=0;
      k = 1 + safe_cat(buf, name, 300);
      byte_copy(buf + k, PROCESS_SIZE, &root[j]);
      k += PROCESS_SIZE;

      safe_write(k);
      safe_read();
      if (verbose) { msg(fu(j),".\t",name,align_name(name),"done"); }
    }

    if (verbose) errmsg_puts(1,"\n");
    msg("Done!\tTry: nsvc -L");
    return(0);
  }
  goto usage;
  return 0;
}
